README.md
# flutter_pay_u_money

A Flutter plugin for payumoney gateway in the mobile platform. Supports Android only.



# Contributing
**Currently supported for only Android Platform. Not for iOS right now.
If you want to make it more better or want to add more platforms support then feel free to contribute.**


#### Initiate Payment

```    
var _payuData = PayUData(
  ///your merchant ID
  merchantId: '',

  ///your merchant key
  merchantKey: '',

  ///your salt
  salt: '',

  ///product name
  productName: 'blah blah',

  /// custom transaction id
  txnId: 'some1234thing',

  ///optional you can add an hash from server side or you can generate from here
  ///the hash sequence should be => key|txnid|amount|productinfo|firstname|email|udf1|udf2|udf3|udf4|udf5||||||salt;
  //hash: '',

  ///this udfs[User Defined Field] is optional you can add up to 10 if you want extra field to pass
  // udf: ['udf1', 'udf2', 'udf3', 'udf4', 'udf5', 'udf6', 'udf7', 'udf8', 'udf9', 'udf10'],
  amount: '125',
  firstName: 'tester',
  email: 'test@gmail.com',
  phone: '9876543210',
);
```
#### For Sandbox
    
```
 var _flutterPayUMoney = FlutterPayUMoney.test(payUData: _payuData);
```

#### For Live
   
```
 var _flutterPayUMoney = FlutterPayUMoney.production(payUData: _payuData);
```
   
#### To Generate Hash Natively

```
 await _flutterPayUMoney.hashIt();
```
#### Start Payment

```
    await _flutterPayUMoney.pay(
      paymentResponse: (statusCode, data) {
        print('$statusCode :: $data');
        setState(() {
          _resultData = '$statusCode :: $data';
        });
      },
    );
```
#### Payment Response Status Code

Code | Status |
--- | --- | 
200 | Success |
400 | Failure |
401 | Validation Error |
402 | User Payment Cancelled  |

- [See example for more details](https://pub.dev/packages/flutter_pay_u_money/example)


#### Server Code
To generate hash from server side

```php
    <?php 
    header('Content-Type: application/json');
    header('Access-Control-Allow-Origin: *');
    
function preparePayuPayment($params)
{        
    $SS_URL = (isset($params['SS_URL']))? $params['SS_URL'] : '';
    $FF_URL = (isset($params['FF_URL']))? $params['FF_URL'] : '';
    $params['key'] = 'KSXB9Z3J';    
    
    $hashSequence = 'key|txnid|amount|productinfo|firstname|email|udf1|udf2|udf3|udf4|udf5|udf6|udf7|udf8|udf9|udf10';
    
    $formError = 0;
    $txnid = '';
    if(empty($params['txnid']))
    {
      $txnid = substr(hash('sha256', mt_rand() . microtime()), 0, 20);
    }
    else
    {
      $txnid = $params['txnid'];
    }    
    
    $params['txnid'] = $txnid;
    $hash = '';
    
    if(empty($params['hash']) && sizeof($params) > 0)
    {
        if(empty($params['txnid']) || empty($params['amount']) || empty($params['firstname']) || empty($params['email']) || empty($params['phone']) || empty($params['productinfo']))
        {
            $formError = 1;
        }
        else
        {
            $hashVarsSeq = explode('|', $hashSequence);
            $hash_string = '';    
            foreach($hashVarsSeq as $hash_var)
            {
                  $hash_string .= isset($params[$hash_var]) ? $params[$hash_var] : '';
                  $hash_string .= '|';
              }
            $hash_string .= $params['SALT_KEY'];
            $hash = strtolower(hash('sha512', $hash_string));
        }
    }
    else if(!empty($params['hash']))
    {
        $hash = $params['hash'];
    }
    
    $params['hash'] = $hash;
    return array('status'=>$formError, 'params'=>$params);
}
if($_SERVER['REQUEST_METHOD'] == 'POST') {
  $allParmas = array(
       'key' => 'KSXB9Z3J',
       'txnid' => $_POST['txnid'],
       'amount' => $_POST['amount'],
       'SALT_KEY' => 'AwuZ5FVG4c',
       'productinfo' => $_POST['productinfo'],
       'firstname' => $_POST['firstname'],
       'email' => $_POST['email'],
       'phone'=> $_POST['phone'],
       'udf1' => '',
       'udf2' => '',
       'udf3' => '',
       'udf4' => '',
       'udf5' => '',
       'udf6' => '',
       'udf7' => '',
       'udf8' => '',
       'udf9' => '',
       'udf10' => ''
     );
     $returnData = preparePayuPayment($allParmas);
     echo json_encode($returnData);
}else {

  echo json_encode(array(
    'error' => 'data not passed',

  ));
}
     ?> 
```

### IMPORTANT
1. You have to setup a server for generating Hash for starting payments then you have to call the function for starting payment. OK.
2. All used payment details are fake but you can use maerchant id and maerchant key for testing because these are test keys provided by PayuMoney

#### Issues
If you are having any issues relating to this plugin then you can create a issue at this github repo.



CHANGELOG.md

## 0.5.7
* Fixed minor bugs
## 0.5.6
* Fixed minor bugs
## 0.5.5
* Fixed Payment Failure
## 0.5.4
* Fixed minor bugs
## 0.5.3
* Fixed minor bugs

## 0.5.2+2
* This is just a version bump to republish as 0.5.2+1 was published with some dirty local state.

## 0.5.2+1
* This is just a version bump to republish as 0.5.2 was published with some dirty local state.

## 0.5.2
* Changed README
* Fixed minor bugs

## 0.5.1
* Simplified example
* Fixed minor bugs

## 0.5.0
* Initial release



example/test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility that Flutter provides. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:flutter_pay_u_money_example/main.dart';

void main() {
  testWidgets('Verify Platform version', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(MyApp());

    // Verify that platform version is retrieved.
    expect(
      find.byWidgetPredicate(
        (Widget widget) =>
            widget is Text && widget.data.startsWith('Running on:'),
      ),
      findsOneWidget,
    );
  });
}


test/flutter_pay_u_money_test.dart
import 'package:flutter/services.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:flutter_pay_u_money/flutter_pay_u_money.dart';

void main() {
  const MethodChannel channel = MethodChannel('flutter_pay_u_money');

  TestWidgetsFlutterBinding.ensureInitialized();

  setUp(() {
    channel.setMockMethodCallHandler((MethodCall methodCall) async {
      return '42';
    });
  });

  tearDown(() {
    channel.setMockMethodCallHandler(null);
  });

  /*test('getPlatformVersion', () async {
    expect(await FlutterPayUMoney.platformVersion, '42');
  });*/
}


lib/pay_u_data.dart
import 'package:flutter/material.dart';

class PayUData {
  PayUData({
    @required this.merchantKey,
    @required this.merchantId,
    @required this.salt,
    this.txnId,
    this.firstName,
    this.email,
    this.phone,
    this.productName,
    this.amount,
    this.hash,
    this.sUrl,
    this.fUrl,
    List<String> udf,
  }) : udf = udf ?? List<String>.generate(10, (index) => '');

  /// [merchantId], [merchantKey] and [salt] will only get when registered in payUmoney
  /// [payu.in](https://www.payu.in/)
  ///
  /// Give merchantId from payumoney.
  String merchantId;

  /// Give merchantKey from payumoney.
  String merchantKey;

  ///There is no need of specific txnId you can provide custom alphanumeric.
  String txnId;

  ///Provide customer name.
  String firstName;

  ///Provide customer email.
  String email;

  ///Provide customer phone.
  String phone;

  ///Provide product name.
  String productName;

  ///Amount you want to pay.
  String amount;

  ///If you generate hash from server provide in this field. Otherwise keep it blank.
  ///
  /// For generating hash natively call.
  /// ```
  /// await flutterPayUMoney.hashIt();
  /// ```
  String hash;

  /// Give salt from payumoney.
  String salt;

  /// success url optional.
  /// default url is [https://www.payumoney.com/mobileapp/payumoney/success.php](https://www.payumoney.com/mobileapp/payumoney/success.php)
  String sUrl;

  /// failure url optional.
  /// default url is [https://www.payumoney.com/mobileapp/payumoney/failure.php](https://www.payumoney.com/mobileapp/payumoney/failure.php)
  String fUrl;

  ///User can add extra field to pass from your app to your server. If nothing leave it blank.
  ///maximum 10 udf
  List<String> udf;

  ///Convert Object to Map
  Map<String, dynamic> toJson() => {
        'merchantKey': merchantKey,
        'merchantId': merchantId,
        'firstName': firstName,
        'email': email,
        'phone': phone,
        'txnId': txnId,
        'productName': productName,
        'amount': amount,
        'hash': hash,
        'salt': salt,
        'sUrl': sUrl,
        'fUrl': fUrl,
        'udf': udf == null ? null : List<dynamic>.from(udf.map((x) => x)),
      };

  @override
  String toString() {
    return 'PayUData{amount: $amount, txnId: $txnId, phone: $phone, merchantKey: $merchantKey, merchantId: $merchantId, '
        'productName: $productName, firstName: $firstName, email: $email, hash: $hash, salt: $salt, udf: $udf}';
  }
}


lib/flutter_pay_u_money.dart
import 'package:flutter/services.dart';
import 'package:flutter/widgets.dart';
import 'package:flutter_pay_u_money/pay_u_data.dart';

final MethodChannel _channel = MethodChannel('flutter_pay_u_money');

class FlutterPayUMoney {
  ///whether it is test or live.
  final bool _isTest;

  ///Object data.
  final PayUData payUData;

  ///For sandbox use test method or use production method
  FlutterPayUMoney.test({
    @required PayUData payUData,
  })  : _isTest = true,
        payUData = payUData,
        assert(payUData != null),
        assert(payUData.udf.length <= 10);

  ///For sandbox use test method or use production method
  FlutterPayUMoney.production({@required PayUData payUData})
      : _isTest = false,
        payUData = payUData,
        assert(payUData != null),
        assert(payUData.udf.length <= 10);

  ///call to launch payumoney gateway.
  ///Before calling this method make sure the hash is not null.
  /// For generating hash natively call.
  /// ```
  /// await flutterPayUMoney.hashIt();
  /// ```
  ///
  /// StatusCode
  /// ```
  /// 200 = Success
  /// 400 = Failed
  /// 401 = Validation Error
  /// 402 = User Payment Cancelled
  /// ```
  Future<void> pay(
      {Function(String statusCode, dynamic) paymentResponse}) async {
    assert(payUData.hash != null && payUData.hash.isNotEmpty,
        'payUData.hash was called on null or empty. generate hash from server side or call method hashIt before calling pay method');

    _generateUDF();
    var result;
    if (_isTest) {
      try {
        result = await _channel
            .invokeMethod('testPay', payUData.toJson())
            .catchError((onError) {
          debugPrint(
              'Payment Failed :: ${(onError as PlatformException).toString()}');
          paymentResponse((onError as PlatformException).code,
              (onError as PlatformException).details);
        });
      } on Exception catch (e) {
        throw 'Something went wrong ${e.toString()}';
      }
    } else {
      try {
        result = await _channel
            .invokeMethod('livePay', payUData.toJson())
            .catchError((onError) {
          debugPrint(
              'Payment Failed :: ${(onError as PlatformException).toString()}');
          paymentResponse((onError as PlatformException).code,
              (onError as PlatformException).details);
        });
      } on Exception catch (e) {
        throw 'Something went wrong ${e.toString()}';
      }
    }
    debugPrint('transaction result :: $result');
    if (result != null) {
      paymentResponse('200', result);
    }
  }

  ///To generate hash natively.
  Future<void> hashIt() async {
    debugPrint(
        'hashData :: ${payUData.merchantKey}|${payUData.txnId}|${payUData.amount}|${payUData.productName}|'
        "${payUData.firstName}|${payUData.email}|${_generateUDF().join("|")}|${payUData.salt}");

    try {
      payUData.hash = await _channel
          .invokeMethod(
              'hashIt',
              '${payUData.merchantKey}|${payUData.txnId}|${payUData.amount}|${payUData.productName}|${payUData.firstName}|'
                  "${payUData.email}|${_generateUDF().join("|")}|${payUData.salt}")
          .catchError((onError) {
        throw (onError as PlatformException).toString();
      });
    } on Exception catch (e) {
      throw 'Hashing error ${e.toString()}';
    }

    debugPrint('hashedIt:: ${payUData.hash}');
  }

  ///Generate UDF up to ten when user gives blank udf.
  List<String> _generateUDF() {
    var createUDF = payUData.udf;
    if (createUDF.length < 10) {
      for (var i = payUData.udf.length; i < 10; i++) {
        createUDF.add('');
      }
    }
    return createUDF;
  }
}


android/src/main/java/com/dithesh/flutter_pay_u_money/FlutterPayUMoneyPlugin.java
package com.dithesh.flutter_pay_u_money;

import android.app.Activity;
import android.content.Intent;

import androidx.annotation.NonNull;

import com.payumoney.core.PayUmoneySdkInitializer;
import com.payumoney.core.entity.TransactionResponse;
import com.payumoney.sdkui.ui.utils.PayUmoneyFlowManager;
import com.payumoney.sdkui.ui.utils.ResultModel;

import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.List;

import io.flutter.Log;
import io.flutter.embedding.engine.plugins.FlutterPlugin;
import io.flutter.embedding.engine.plugins.activity.ActivityAware;
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;
import io.flutter.plugin.common.PluginRegistry;

import static android.app.Activity.RESULT_OK;

/**
 * FlutterPayUMoneyPlugin
 */
public class FlutterPayUMoneyPlugin implements FlutterPlugin, MethodCallHandler, ActivityAware, PluginRegistry.ActivityResultListener {
    private static final String TAG = "FlutterPayUMoneyPlugin";
    private MethodChannel channel;
    private Activity activity;
    private MethodChannel.Result wholeResult;

    @Override
    public void onAttachedToEngine(@NonNull FlutterPluginBinding flutterPluginBinding) {
        channel = new MethodChannel(flutterPluginBinding.getBinaryMessenger(), "flutter_pay_u_money");
        channel.setMethodCallHandler(this);
    }

    @Override
    public void onMethodCall(@NonNull MethodCall call, @NonNull Result result) {
        switch (call.method) {
            case "hashIt":
                generateHashKey(String.valueOf(call.arguments), result);
                break;
            case "testPay":
                System.out.println("arguments :: " + call.arguments);
                wholeResult = result;
                launchPayUMoneyFlow(call, true);
                break;
            case "livePay":
                System.out.println("arguments :: " + call.arguments);
                wholeResult = result;
                launchPayUMoneyFlow(call, false);
                break;
            default:
                result.notImplemented();
                break;
        }
    }


    @Override
    public void onDetachedFromEngine(@NonNull FlutterPluginBinding binding) {
        channel.setMethodCallHandler(null);
    }


    @Override
    public void onAttachedToActivity(ActivityPluginBinding binding) {
        activity = binding.getActivity();
        binding.addActivityResultListener(this);

    }

    @Override
    public void onDetachedFromActivityForConfigChanges() {

    }

    @Override
    public void onReattachedToActivityForConfigChanges(@NonNull ActivityPluginBinding binding) {

    }

    @Override
    public void onDetachedFromActivity() {

    }


    private void launchPayUMoneyFlow(MethodCall call, boolean environment) {
        String udf1 = "",
                udf2 = "",
                udf3 = "",
                udf4 = "",
                udf5 = "",
                udf6 = "",
                udf7 = "",
                udf8 = "",
                udf9 = "",
                udf10 = "";
        PayUmoneySdkInitializer.PaymentParam.Builder builder = new PayUmoneySdkInitializer.PaymentParam.Builder();
        List<String> udfs = call.argument("udf");
        String merchantKey = call.argument("merchantKey");
        String merchantId = call.argument("merchantId");
        String sUrl = call.argument("sUrl") != null ? (String) call.argument("sUrl") : "https ://www.payumoney.com/mobileapp/payumoney/success.php";
        String fUrl = call.argument("fUrl") != null ? (String) call.argument("fUrl") : "https://www.payumoney.com/mobileapp/payumoney/failure.php";
        String firstName = call.argument("firstName");
        String email = call.argument("email");
        String phone = call.argument("phone");
        String productName = call.argument("productName");
        String amount = call.argument("amount");
        String txnId = call.argument("txnId");
        if (udfs != null) {
            udf1 = udfs.get(0);
            udf2 = udfs.get(1);
            udf3 = udfs.get(2);
            udf4 = udfs.get(3);
            udf5 = udfs.get(4);
            udf6 = udfs.get(5);
            udf7 = udfs.get(6);
            udf8 = udfs.get(7);
            udf9 = udfs.get(8);
            udf10 = udfs.get(9);
        }
        String hash = call.argument("hash");

        builder.setAmount(amount)
                .setTxnId(txnId)
                .setPhone(phone)
                .setProductName(productName)
                .setFirstName(firstName)
                .setEmail(email)
                .setIsDebug(environment)
                .setsUrl(sUrl)
                .setfUrl(fUrl)
                .setUdf1(udf1)
                .setUdf2(udf2)
                .setUdf3(udf3)
                .setUdf4(udf4)
                .setUdf5(udf5)
                .setUdf6(udf6)
                .setUdf7(udf7)
                .setUdf8(udf8)
                .setUdf9(udf9)
                .setUdf10(udf10)
                .setKey(merchantKey)
                .setMerchantId(merchantId);

        try {
            PayUmoneySdkInitializer.PaymentParam mPaymentParams = builder.build();
            mPaymentParams.setMerchantHash(hash);
            PayUmoneyFlowManager.startPayUMoneyFlow(mPaymentParams, activity, R.style.AppTheme_default, false);
        } catch (Exception e) {
            wholeResult.error("401", e.getMessage(), e.getMessage());
            e.printStackTrace();

        }
    }


    // String hashSequence = "key|txnid|amount|productinfo|firstname|email|udf1|udf2|udf3|udf4|udf5||||||salt";
    private void generateHashKey(String hashString, Result result) {
        StringBuilder hash = new StringBuilder();
        MessageDigest messageDigest;
        try {
            messageDigest = MessageDigest.getInstance("SHA-512");
            messageDigest.update(hashString.getBytes());
            byte[] mdbytes = messageDigest.digest();
            for (byte hashByte : mdbytes) {
                hash.append(Integer.toString((hashByte & 0xff) + 0x100, 16).substring(1));
            }
            result.success(hash.toString());
        } catch (NoSuchAlgorithmException e) {
            result.error("400", "hashing error", e);
            e.printStackTrace();

        }
    }

    @Override
    public boolean onActivityResult(int requestCode, int resultCode, Intent data) {
        // Result Code is -1 send from Payumoney activity
        Log.d(FlutterPayUMoneyPlugin.TAG, "request code " + requestCode + " resultcode " + resultCode);
        if (requestCode == PayUmoneyFlowManager.REQUEST_CODE_PAYMENT && resultCode == RESULT_OK && data !=
                null) {
            TransactionResponse transactionResponse = data.getParcelableExtra(PayUmoneyFlowManager
                    .INTENT_EXTRA_TRANSACTION_RESPONSE);

            ResultModel resultModel = data.getParcelableExtra(PayUmoneyFlowManager.ARG_RESULT);
            // Check which object is non-null
            if (transactionResponse != null && transactionResponse.getPayuResponse() != null) {
                if (transactionResponse.getTransactionStatus().equals(TransactionResponse.TransactionStatus.SUCCESSFUL)) {
                    //Success Transaction
                    wholeResult.success(transactionResponse.getPayuResponse());
                } else {
                    //Failure Transaction
                    wholeResult.error("400", "Payment Failed", transactionResponse.getPayuResponse());
                }

            } else if (resultModel != null && resultModel.getError() != null) {
                Log.d(TAG, "Error response : " + resultModel.getError().getTransactionResponse());
            } else {
                Log.d(TAG, "Both objects are null!");
            }
        } else {
            if (requestCode == PayUmoneyFlowManager.REQUEST_CODE_PAYMENT)
                wholeResult.error("402", "User Cancelled Payment", "User Cancelled Payment");
        }
        return true;
    }
}


example/lib/main.dart
import 'dart:async';

import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';
import 'package:flutter_pay_u_money/flutter_pay_u_money.dart';
import 'package:flutter_pay_u_money/pay_u_data.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatefulWidget {
  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  var _resultData;

  @override
  void initState() {
    super.initState();
  }

  Future<void> initPlatformState() async {
    var _payuData = PayUData(
      ///your merchant ID
      merchantId: '',

      ///your merchant key
      merchantKey: '',

      ///your salt
      salt: '',

      ///product name
      productName: 'blah blah',

      /// custom transaction id
      txnId: 'some1234thing',
      amount: '125',
      firstName: 'tester',
      email: 'test@gmail.com',
      phone: '9876543210',

      ///optional you can add an hash from server side or you can generate from here
      ///the hash sequence should be => key|txnid|amount|productinfo|firstname|email|udf1|udf2|udf3|udf4|udf5||||||salt;
      // hash: '',

      ///this udf [User Defined Field] is optional you can add up to 10 if you want extra field to pass
      udf: [
        'udf1',
        'udf2',
        'udf3',
        'udf4',
        'udf5',
        'udf6',
        'udf7',
        'udf8',
        'udf9',
        'udf10',
      ],
      // sUrl: '',
      // fUrl: '',
    );

    ///for sandbox
    var _flutterPayUMoney = FlutterPayUMoney.test(payUData: _payuData);

    ///for live
    // var _flutterPayUMoney = FlutterPayUMoney.production(payUData: _payuData);

    ///if you want generate hash call [hashIt()]
    await _flutterPayUMoney.hashIt();
    await _flutterPayUMoney.pay(
      paymentResponse: (statusCode, data) {
        print('$statusCode :: $data');
        setState(() {
          _resultData = '$statusCode :: $data';
        });
      },
    );
  }

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(
          title: const Text('Flutter PayUMoney example app'),
        ),
        body: Column(
          crossAxisAlignment: CrossAxisAlignment.center,
          mainAxisAlignment: MainAxisAlignment.center,
          children: [
            Center(
              child: RaisedButton(
                  onPressed: () {
                    initPlatformState();
                  },
                  child: Text('Pay')),
            ),
            Text('Result: $_resultData'),
          ],
        ),
      ),
    );
  }
}
